finetune/base_models/bert/featurizer.py

```diff
@@ -140,4 +145,279 @@ def bert_featurizer(
+        # Appears to be 1 where you have tokens and 0 where you have padding
+        # [batch_size, sequence_length]
+        mask = tf.sequence_mask(lengths, maxlen=seq_length, dtype=tf.float32)
+        bert = underlying_model(
+            config=bert_config,
+            is_training=train,
+            input_ids=X_sub,
+            input_mask=mask,
+            input_context=context,
+            token_type_ids=token_type_ids,
+            use_one_hot_embeddings=False,
+            scope=None,
+            use_pooler=config.bert_use_pooler,
+            use_token_type=config.bert_use_type_embed,
+            reading_order_decay_rate=reading_order_decay_rate,
+        )
+        return bert.get_pooled_output()
+
+    def chunk_aggregation(pooled_outputs):
+        """
+        Function for aggregating/pooling the pooled outputs from the BERT featurizer
+        across the chunk dimension
+
+        TODO Implement max, attention, lstm, and concat options
+        """
+        if config.chunk_pool_fn == "mean":
+            aggr_outputs = tf.reduce_mean(pooled_outputs, axis=0)
+        elif config.chunk_pool_fn == "attention":
+            aggr_outputs = attn(pooled_outputs)
+        # This is wrong
+        else:
+            aggr_outputs = tf.reduce_mean(pooled_outputs, axis=0)
+
+        return aggr_outputs
+
+    with tf.compat.v1.variable_scope("model/featurizer", reuse=reuse):
+        # Get pooled output across chunk dim using map_fn
+        # [num_chunks / batch_scaling, batch_size * batch_scaling, hidden_size]
+        # pooled_output = tf.map_fn(lambda inp: bert_wrapper(inp[0], inp[1], inp[2]), (X, mask, token_type_ids))
+        pooled_output = tf.map_fn(lambda inp: bert_wrapper(inp), X, dtype=tf.float32)
+        pooled_output = tf.compat.v1.Print(
+            pooled_output, ["pooled_output", tf.shape(pooled_output)], summarize=-1
+        )
+
+        # Reshape output to remove batch_scaling factor
+        # [num_chunks, batch_size, hidden_size]
+        pool_shape = tf.shape(input=pooled_output)
+        pooled_output = tf.reshape(
+            pooled_output,
+            [
+                pool_shape[0] * config.batch_size_scaler,
+                tf.cast(pool_shape[1] / config.batch_size_scaler, dtype=tf.int32),
+                pool_shape[2],
+            ],
+        )
+        pooled_output = tf.compat.v1.Print(
+            pooled_output, ["pooled_rs", tf.shape(pooled_output)], summarize=-1
+        )
+
+        # Reduce across chunk dim with aggregation operation
+        # TODO implement aggregation options other than mean and attention
+        features = chunk_aggregation(pooled_output)
+        features = tf.compat.v1.Print(
+            features, ["features", tf.shape(features)], summarize=-1
+        )
+
+        # FIXME In original featurizer, this reshape appears not to change dims [batch_size, hidden_size],
+        # but without it there are shape errors in the target model
```

Comment:
Could mean you just need to .set_shape()? Not sure though

Reply:
The featurizers can accept an arbitrary number of batch dims that get flattened out and then reintroduced later - this is used for Comparison etc. so this does need to be a reshape, even though for most cases it will effectively just set the shape.